File: java/linkedlist/ReverseLinkedList.java
package linkedlist;

public class ReverseLinkedList {
    /* 可以用for loop翻转 细节在于保持一个previous node（一开始dummy）和当前node进行交换
     * 可以用递归实现 如果我们假设reverseList能够成功 那么其会返回翻转后的head 也就是翻转前的last
     * 而原先的head next 现在其next指向null 但是head的next仍然指向head next
     * 所以根据这个情况 我们只要把head next的next从null修改成head 然后将head的next修改成null就可以了
     * 递归的停止条件是null或者next是null 如果next是null的话 翻转就是自己做head 可以直接return自己
     */
    public ListNode reverseList(ListNode head) {
        if(head == null || head.next == null) {
            return head;
        }
        ListNode last = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return last;
    }
}

File: c++/linkedlist/reverseLinkedList.cpp
#include <include/listnode.h>
/**
 * 206. Reverse Linked List
 * C++版本 思路非常简单 细节上牢记有四步 并且利用递归特质 不要设想在一个节点上
 * 翻转完自己、前者和后者, 每次只动自己的next, 剩下的不过是为这个目的服务罢了
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr, *next;
        while(head) {
            next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        return prev; // head would be null
    }
};